### PR TITLE
FEAT- AAVE stoploss

### PR DESCRIPTION
**Pull request by halaprix**

### CODE REVIEW

File: contracts/commands/AaveStoplLossCommand.sol [hunk 1]
```diff
@@ -0,0 +1,307 @@
...
+            CloseData memory closeData;
+
+            address[] memory debtTokens = new address[](1);
+            debtTokens[0] = address(aaveData.debtTokenAddress);
+            uint256[] memory amounts = new uint256[](1);
+            amounts[0] = (101 * totalToRepay) / 100;
+            uint256[] memory modes = new uint256[](1);
+            modes[0] = uint256(0);
+
+            closeData.receiverAddress = addressRegistry.aaveStopLoss;
```
Comment (adamskrodzki) [inline contracts/commands/AaveStoplLossCommand.sol:283]: self

File: contracts/commands/AaveStoplLossCommand.sol [hunk 2]
```diff
@@ -0,0 +1,307 @@
...
+            );
+            closeData.referralCode = 0;
+            lendingPool.flashLoan(
+                closeData.receiverAddress,
+                closeData.assets,
+                closeData.amounts,
+                closeData.modes,
+                closeData.onBehalfOf,
+                closeData.params,
+                closeData.referralCode
```
Comment (adamskrodzki) [inline contracts/commands/AaveStoplLossCommand.sol:303]: just 0 ?

File: contracts/commands/AaveStoplLossCommand.sol [hunk 3]
```diff
@@ -0,0 +1,307 @@
...
+            "coll token in command",
+            IERC20(collateralTokenAddress).balanceOf(address(this))
+        );
+        // approve swap to be able to transfer out the collateral token - TODO change max to actual aToken/coll token  balance
+        collateralToken.approve(exchangeAddress, type(uint256).max);
+
+        uint256 paybackReceivedFromSwap = ISwap(exchangeAddress).swapTokens(exchangeData);
+        console.log("returned from swap", paybackReceivedFromSwap);
+        require(
+            paybackReceivedFromSwap > (flData.amounts[0] + flData.premiums[0]),
```
Comment (adamskrodzki) [inline contracts/commands/AaveStoplLossCommand.sol:165]: there is no receiveAtLeast value  ? if there is then use this one 

You can also check that receiveAtLeast >= flData.amounts[0] + flData.premiums[0]

File: contracts/commands/AaveStoplLossCommand.sol [hunk 4]
```diff
@@ -0,0 +1,307 @@
...
+            "aapa/recieved-too-little-from-swap"
+        );
+        uint256 usdBalance = debtToken.balanceOf(address(this));
+        // send remaining debtToken to proxy TODO - send to user
+        debtToken.transfer(borrower, usdBalance - (flData.amounts[0] + flData.premiums[0]));
```
Comment (adamskrodzki) [inline contracts/commands/AaveStoplLossCommand.sol:170]: last time I've checked debt tokens where non-transferable

File: contracts/commands/AaveStoplLossCommand.sol [hunk 5]
```diff
@@ -0,0 +1,307 @@
...
+        console.log("| amount", flData.amounts[0]);
+        console.log("| payback amount", flData.amounts[0] + flData.premiums[0]);
+        console.log("X----------------------");
+
+        require(msg.sender == address(lendingPool), "aaveSl/caller-must-be-lending-pool");
+        console.log("flData.onBehalfOf", flData.onBehalfOf);
+        console.log("initiator", initiator);
+
+        // FL token / debt token
+        address debtTokenAddress = assets[0];
```
Comment (adamskrodzki) [inline contracts/commands/AaveStoplLossCommand.sol:131]: names are mixing, with AAVE debt token,

File: contracts/commands/AaveStoplLossCommand.sol (hunk 4 above)
Comment (adamskrodzki) [reply contracts/commands/AaveStoplLossCommand.sol:170]: NVM, just bad name

**Review by adamskrodzki — CHANGES_REQUESTED** [review]

File: contracts/commands/AaveStoplLossCommand.sol (hunk 3 above)
Comment (halaprix) [reply contracts/commands/AaveStoplLossCommand.sol:165]: it's only in the exchange data that we pass directly to swap contract. I think we are fine here

File: contracts/commands/AaveStoplLossCommand.sol (hunk 2 above)
Comment (halaprix) [reply contracts/commands/AaveStoplLossCommand.sol:303]:
```
The referral program is currently in active and you can pass0 as thereferralCode.
In future for referral code to be active again, a governance proposal, with the list of unique referral codes for various integration must be passed via governance.
```

https://docs.aave.com/developers/v/2.0/the-core-protocol/lendingpool#deposit

File: contracts/commands/AaveStoplLossCommand.sol [hunk 6]
```diff
@@ -0,0 +1,307 @@
...
+        );
+        uint256 totalDebt = IERC20(collReserveData.aTokenAddress).balanceOf(aaveData.borrower);
+        IERC20(collReserveData.aTokenAddress).approve(addressRegistry.aaveStopLoss, totalDebt);
+        {
+            CloseData memory closeData;
+
+            address[] memory debtTokens = new address[](1);
+            debtTokens[0] = address(aaveData.debtTokenAddress);
+            uint256[] memory amounts = new uint256[](1);
+            amounts[0] = (101 * totalToRepay) / 100;
```
[the remark this answers is not in the document]
Comment (halaprix) [reply contracts/commands/AaveStoplLossCommand.sol:279]: yes. that's a remiander of some previous testing. reverted to `totalToRepay`

File: contracts/commands/AaveStoplLossCommand.sol [hunk 7]
```diff
@@ -0,0 +1,307 @@
...
+            (StopLossTriggerData)
+        );
+
+        // TODO legit caller write here and check in closePosition
+        console.log("msg sender", msg.sender);
+        console.log("command execute address this", address(this));
+        trustedCaller = stopLossTriggerData.positionAddress;
+        IAccountImplementation(stopLossTriggerData.positionAddress).execute(
+            address(this),
+            executionData
```
[the remark this answers is not in the document]
Comment (halaprix) [reply contracts/commands/AaveStoplLossCommand.sol:237]: ok done

File: contracts/commands/AaveStoplLossCommand.sol [hunk 8]
```diff
@@ -0,0 +1,307 @@
...
+            stopLossTriggerData.slLevel > 100 &&
+            (stopLossTriggerData.triggerType == 10 || stopLossTriggerData.triggerType == 11);
+    }
+
+    function closePosition(
+        EarnSwapData.SwapData calldata exchangeData,
+        AaveData memory aaveData,
+        AddressRegistry calldata addressRegistry
+    ) external {
+        DataTypes.ReserveData memory collReserveData = lendingPool.getReserveData(
```
[the remark this answers is not in the document]
Comment (halaprix) [reply contracts/commands/AaveStoplLossCommand.sol:262]: yes but it's in proxy context so
```
      require(
            AaveStopLoss(addressRegistry.aaveStopLoss).self() == msg.sender,
            "aaveSl/msg-sender-is-not-sl"
        );
```

File: contracts/commands/AaveStoplLossCommand.sol [hunk 9]
```diff
@@ -0,0 +1,307 @@
...
+            aaveData.collateralTokenAddress
+        );
+        DataTypes.ReserveData memory debtReserveData = lendingPool.getReserveData(
+            aaveData.debtTokenAddress
+        );
+        uint256 totalToRepay = IERC20(debtReserveData.variableDebtTokenAddress).balanceOf(
+            aaveData.borrower
+        );
+        uint256 totalDebt = IERC20(collReserveData.aTokenAddress).balanceOf(aaveData.borrower);
```
[the remark this answers is not in the document]
Comment (halaprix) [reply contracts/commands/AaveStoplLossCommand.sol:271]: of course :)

**Review by adamskrodzki — APPROVED** [review]

**Pull request merged**